kratos/geometries/nurbs_curve_geometry.h

```diff
@@ -0,0 +1,371 @@
+    typedef NurbsCurveGeometry<TWorkingSpaceDimension, TPointType> GeometryType;
+
+    typedef typename BaseType::IndexType IndexType;
+    typedef typename BaseType::SizeType SizeType;
+
+    /** Array of counted pointers to point. This type used to hold
+        geometry's points.*/
+    typedef  typename BaseType::PointsArrayType PointsArrayType;
+
+    /// Counted pointer of NurbsSurfaceShapeFunction
+    KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION(NurbsCurveGeometry);
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /* Conctructor for B-Spline curves. */
+    NurbsCurveGeometry(
+        const PointsArrayType& rThisPoints,
+        const int PolynomialDegree,
+        const Vector& rKnots)
+        : BaseType(PointsArrayType(), &msGeometryData)
+        , mPolynomialDegree(PolynomialDegree)
+        , mKnots(rKnots)
+        , mIsRational(false)
+    {
+        KRATOS_DEBUG_ERROR_IF(rKnots.size() != NurbsUtilities::GetNbKnots(PolynomialDegree, rThisPoints.size()))
+            << "Number of knots and control points do not match!" << std::endl;
+    }
+
+    /* Conctructor for NURBS curves. */
+    NurbsCurveGeometry(
+        const PointsArrayType& rThisPoints,
+        const int PolynomialDegree,
+        const Vector& rKnots,
+        const Vector& rWeights)
+        : BaseType(PointsArrayType(), &msGeometryData)
+        , mPolynomialDegree(PolynomialDegree)
+        , mKnots(rKnots)
+        , mIsRational(true)
+        , mWeights(rWeights)
+    {
+        KRATOS_DEBUG_ERROR_IF(rKnots.size() != NurbsUtilities::GetNumberOfKnots(PolynomialDegree, rThisPoints.size()))
+            << "Number of knots and control points do not match!" << std::endl;
+
+        KRATOS_DEBUG_ERROR_IF(rWeights.size() != rThisPoints.size())
+            << "Number of control points and weights do not match!" << std::endl;
+    }
+
+    explicit NurbsCurveGeometry(const PointsArrayType& ThisPoints)
+        : BaseType(ThisPoints, &msGeometryData)
+    {
+    }
+
+    /* Copy constructor.*/
+    NurbsCurveGeometry(NurbsCurveGeometry const& rOther)
+        : BaseType(rOther)
+    {
+    }
+
+    /* Copy constructor from a geometry with different point type.*/
+    template<class TOtherPointType> NurbsCurveGeometry(
+        NurbsCurveGeometry<TWorkingSpaceDimension, TOtherPointType> const& rOther)
+        : BaseType(rOther)
+    {
+    }
+
+    /* Destructor.*/
+    ~NurbsCurveGeometry() override {}
+
+    GeometryData::KratosGeometryFamily GetGeometryFamily() const override
+    {
+        return GeometryData::Kratos_generic_family;
+    }
+
+    GeometryData::KratosGeometryType GetGeometryType() const override
+    {
+        return GeometryData::Kratos_generic_type;
+    }
+
+    ///@}
+    ///@name Get and Set functions
+    ///@{
+
+    bool IsRational() const
+    {
+        return mWeights.size() != 0;
+    }
+
+    const Vector& Knots() const
+    {
+        return mKnots;
+    }
+
+    int NumberOfKnots() const
```

Comment:
Why int and not std::size_t?

Reply:
The reason for the ints: The algorithms in the nurbs book assume that signed-integer operations can be performed. In some cases usigned integers led to wrong results. Since no error occurs, it is very difficult to debug something like this. Adapting the algorithms from the book is also not very easy in these cases.